### PR TITLE
check next_execution_date and prev_execution_date None

### PR DESCRIPTION
**Pull request by milton0825**
We should check if next_execution_date and prev_execution_date is None.

Found this bug in the  upstream CI...

### CODE REVIEW

**Comment by ArgentFalcon** [comment]
`None` is not a good default, especially if as a user, I use 
{{ next_ds }}

We should use a value that's still valid in this context.

File: airflow/models.py [hunk 1]
```diff
@@ -1494,8 +1494,13 @@ def get_template_context(self, session=None):
         prev_execution_date = task.dag.previous_schedule(self.execution_date)
         next_execution_date = task.dag.following_schedule(self.execution_date)
 
-        next_ds = next_execution_date.strftime('%Y-%m-%d')
-        prev_ds = prev_execution_date.strftime('%Y-%m-%d')
+        next_ds = None
```
Comment (feng-tao) [inline airflow/models.py]: about just set it to tomorrow_ds

File: airflow/models.py [hunk 2]
```diff
@@ -1494,8 +1494,13 @@ def get_template_context(self, session=None):
...
+        if next_execution_date:
+            next_ds = next_execution_date.strftime('%Y-%m-%d')
+
+        prev_ds = None
```
Comment (feng-tao) [inline airflow/models.py]: just set it to yesterday_ds

File: airflow/models.py (hunk 1 above)
Comment (ArgentFalcon) [reply airflow/models.py]: Would sort of work, but could break assumption /shrug

Probably need to update the docs accordingly then.

Comment (milton0825) [reply airflow/models.py]: another way is to actually ensure that next_execution_date is always not None

**Comment by milton0825** [comment]
@ArgentFalcon @feng-tao 
{{ next_ds }} is just {{ next_execution_date }} in `YYYY-MM-DD` format.
We need to make sure the behavior of {{ next_ds }} aligns with {{ next_execution_date }}.

So I think we should do either:
option 1:
If {{ next_execution_date }} is None, we set {{ next_ds }} to None.

option 2:
We make sure {{ next_execution_date }} is never None.

**Comment by ArgentFalcon** [comment]
Hmm, good point. this sounds like a good question for the mailing list?

**Comment by astahlman** [comment]
Option 1 makes the most sense to me. Not sure I understand why we're trying to avoid `None`

**Comment by ArgentFalcon** [comment]
Because returning `None` will return the jinja template result `"None"` which is not super useful when used in SQL. An empty string might be superior.

**Comment by feng-tao** [comment]
per discussion: option 1 seems better(dag could have @once schedule which won't prev / next_schedule)

**Comment by milton0825** [comment]
docs added.

File: docs/code.rst [hunk 3]
```diff
@@ -120,7 +120,9 @@ Variable                            Description
 =================================   ====================================
 ``{{ ds }}``                        the execution date as ``YYYY-MM-DD``
 ``{{ prev_ds }}``                   the previous execution date as ``YYYY-MM-DD``
+                                    and it resolves to ``None`` when ``schedule_interval="@once"``
```
Comment (astahlman) [inline docs/code.rst]: Or if there is no previous execution date, right? e.g., the first DagRun will have no previous execution date. Maybe we should just say something like `the previous execution date as ``YYYY-MM-DD`` if it exists, else None`?

Comment (milton0825) [reply docs/code.rst]: That's better.

Comment (milton0825) [reply docs/code.rst]: addressed.

**Comment by milton0825** [comment]
updated.

File: docs/code.rst [hunk 4]
```diff
@@ -166,6 +166,10 @@ UI. You can access them as either plain-text or JSON. If you use JSON, you are
 also able to walk nested structures, such as dictionaries like:
 ``{{ var.json.my_dict_var.key1 }}``
 
+You can use Jinja filters like
+``{{ prev_ds|default('prev_ds is not defined') }}``
```
Comment (ArgentFalcon) [inline docs/code.rst]: Nice 👍 
I was looking for the filters, where did you find them?

**Comment by ArgentFalcon** [comment]
👍

**Pull request merged**